File: src/main/java/com/alxan/noteefy/event/Event.java
package com.alxan.noteefy.event;

import com.alxan.noteefy.common.BaseIdentifiable;

import java.util.Set;

public class Event<T> extends BaseIdentifiable implements Taggable, Publishable, Comparable<Event<?>> {
    private final T content;
    private final Class<T> contentType;
    private final EventOrder eventOrder = new EventOrder();
    private final TagRegistry tagRegistry = new TagRegistry();

    public Event(T aContent) {
        contentType = (Class<T>) aContent.getClass(); // TODO Fix this
        content = aContent;
    }

    public static <T> Event<?> createEventIfNecessary(T content) {
        if (content instanceof Event) {
            return (Event<?>) content;
        } else {
            return new Event<T>(content);
        }
    }

    public T getContent() {
        return content;
    }

    public Class<T> getContentType() {
        return contentType;
    }

    public Long getOrder() {
        return eventOrder.getOrder();
    }

    @Override
    public int compareTo(Event<?> other) {
        return getOrder().compareTo(other.getOrder());
    }

    @Override
    public Set<String> getTags() {
        return tagRegistry.getTags();
    }
}

File: src/event/Event.kt
package event

import publish.base.Publishable
import java.time.Instant
import java.util.*
import kotlin.collections.HashSet

open class Event<out T : Any>(
    private val clazz: Class<T>,
    private val content: T? = null,
    private val tags: HashSet<String> = hashSetOf()
) : Publishable, Comparable<Event<Any>> {

    private var uuid = UUID.randomUUID()
    private val order = Instant.now().toEpochMilli()
    override fun getId(): UUID = uuid
    override fun getClazz(): Class<*> = clazz
    override fun getOrder(): Long = order
    fun getContent(): T? = content
    fun getTags() = tags.toSet()
    fun hasTag(tag: String) = tags.contains(tag)
    fun addTag(tag: String) = tags.add(tag)
    override fun compareTo(other: Event<Any>): Int = order.compareTo(other.order)
    override fun toString(): String {
        return content?.toString() ?: "Event(${this::class.java.simpleName}, $uuid)"
    }

    companion object {
        fun <T : Any> from(content: T, tags: Set<String> = setOf()): Event<T> {
            return Event(content.javaClass, content, tags.toHashSet())
        }
    }
}